elastic/beats
# Create prototype shipper-beat

## What does this PR do?

See https://github.com/elastic/beats/issues/35135 for context

This creates a shipper filebeat input. When run, it starts a single filebeat input and gRPC server per-agent input, configures any processors, then sends the output upstream as any other filebeat input normally would.

The changes needed for this include:

- A new `shipper` filebeat input
- Changes to the V2 central management code so we don't start a shipper input per-datastream
- Changes to the fileset V2 Client API, so `Publish()` returns a `queue.EntryID` value.
- Changes to the fileset V2 Pipeline API, so it exposes a `PersistedIndex()` call.
- Changes to the various queues to add a `PersistedIndex()` call. The alternative would have been to let the publisher interfaces just expose the `Metrics()` call already implemented by the various queue types, and have the shipper use the metrics data to calculate PersistedIndex, but the disk queue doesn't support the needed metrics for `PersistedIndex()`, and I didn't want to assume that a future implementation would work the same way in the disk queue as it does in the memory queue, so I added a `PersistedIndex()` call instead.
- Change a dozen different mock testing interfaces across the codebase for the above change

Because of all the changes to various tests and mock interfaces, I suspect CI will be broken for a while I hammer out the details.

## What currently works

- The shipper will start up and configure itself, and configure any per-datastream processors based on an input config
- it will send and receive events in a way that appears correct

## Checklist


## Author's Checklist

- [ ] Add tests - not sure if we want to use existing filebeat testing infrastructure, or try and use some of the new integration test suite tools.

## How to test this PR locally

- Right now, this has currently only been tested with agent in standalone mode.
1. Pull down, build this PR
2. Build or download the latest 8.8.0-SNAPSHOT build of elastic-agent
3. unpack elastic-agent, place a build of filebeat from this branch into elastic-agent's `components` directory
4. Add the following blurb to `filebeat.spec.yml`, under the `inputs` section:
```
  - name: shipper
    description: "Shipper"
    platforms: *platforms
    outputs: *outputs
    shippers: *shippers
    command: *command
```
5. modify the `elastic-agent.yml` config to start a shipper input. Example:
```

inputs:
  - type: shipper
    # Each input must have a unique ID.
    id: shipper-id-one
    data_stream:
      namespace: default
    use_output: default
    server: /tmp/test-shipper-beat.sock
    streams:
      - id: system/metrics-test
        metricsets:
        - process_summary
        data_stream:
          dataset: system.process_summary
          type: metrics
        period: 1m
        processors:
        - drop_event.when.regexp:
            system.filesystem.mount_point: ^/(sys|cgroup|proc|dev|etc|host|lib|snap)($|/)
```
Note that this is just a generic metricbeat input config, but with the `type` changed to `shipper`, and the `server` component added.

6. Start elastic-agent, make sure the filebeat-shipper starts successfully
7. Start another beat that's configured to use the filebeat-shipper as the output. This can be done by setting the output config as such:
```
output.shipper:
  server: unix:///tmp/test-shipper-beat.sock
```
